### PR TITLE
only slice SocketAddress on success operation

### LINKED ISSUE #90125: Socket UDP test failures due to endpoint mismatches
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=365410
Build error leg or test failing: System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected
Pull request: https://github.com/dotnet/runtime/pull/90094

```json
{
  "ErrorMessage": "Actual:   0.0.0.0:0",
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=365410
**Error message validated:** `Actual:   0.0.0.0:0`
**Result validation: :white_check_mark:** Known issue matched with the provided build.
**Validation performed at:** 8/7/2023 11:50:20 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[369570](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369570)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369570&view=ms.vss-test-web.build-test-results-tab&runId=7839958&resultId=104133)||
|[369569](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369569)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369569&view=ms.vss-test-web.build-test-results-tab&runId=7838970&resultId=103496)||
|[369466](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369466)|dotnet/runtime|[System.Net.Sockets.Tests.SendReceiveUdpClient.SendToRecvFromAsync_Datagram_UDP_UdpClient](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369466&view=ms.vss-test-web.build-test-results-tab&runId=7836242&resultId=103793)||
|[369431](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369431)|dotnet/runtime|[System.Net.Sockets.Tests.SendReceiveUdpClient.SendToRecvFromAsync_Datagram_UDP_UdpClient](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369431&view=ms.vss-test-web.build-test-results-tab&runId=7836040&resultId=104163)|dotnet/runtime#90253|
|[369132](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369132)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=369132&view=ms.vss-test-web.build-test-results-tab&runId=7829550&resultId=104133)|dotnet/runtime#90284|
|[367934](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367934)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367934&view=ms.vss-test-web.build-test-results-tab&runId=7792074&resultId=103527)||
|[367932](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367932)|dotnet/runtime|[System.Net.Sockets.Tests.SendReceiveUdpClient.SendToRecvFromAsync_Datagram_UDP_UdpClient](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367932&view=ms.vss-test-web.build-test-results-tab&runId=7793044&resultId=104097)||
|[367930](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367930)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=367930&view=ms.vss-test-web.build-test-results-tab&runId=7792022&resultId=104006)||
|[366452](https://dev.azure.com/dnceng-public/public/_build/results?buildId=366452)|dotnet/runtime|[System.Net.Sockets.Tests.SendReceiveUdpClient.SendToRecvFromAsync_Datagram_UDP_UdpClient](https://dev.azure.com/dnceng-public/public/_build/results?buildId=366452&view=ms.vss-test-web.build-test-results-tab&runId=7752632&resultId=104188)||
|[366447](https://dev.azure.com/dnceng-public/public/_build/results?buildId=366447)|dotnet/runtime|[System.Net.Sockets.Tests.SendReceiveUdpClient.SendToRecvFromAsync_Datagram_UDP_UdpClient](https://dev.azure.com/dnceng-public/public/_build/results?buildId=366447&view=ms.vss-test-web.build-test-results-tab&runId=7752516&resultId=104188)||
|[365410](https://dev.azure.com/dnceng-public/public/_build/results?buildId=365410)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=365410&view=ms.vss-test-web.build-test-results-tab&runId=7730198&resultId=104068)|dotnet/runtime#90094|
|[365214](https://dev.azure.com/dnceng-public/public/_build/results?buildId=365214)|dotnet/runtime|[System.Net.Sockets.Tests.TelemetryTest.EventSource_EventsRaisedAsExpected](https://dev.azure.com/dnceng-public/public/_build/results?buildId=365214&view=ms.vss-test-web.build-test-results-tab&runId=7723894&resultId=104015)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|5|12|12|

### PR DESCRIPTION
**Pull request by wfurt**
This is regression from https://github.com/dotnet/runtime/pull/88970.
When the SocketPal.TryCompleteReceiveFrom fails e.g. no data yet, we should not Slice the SocketAddress because on next round we would not have buffer to receive to.  (as we do in case just bellow)
This was added to avoid tracking SocketAddressLength as separate parameter. 

Change is UpdateReceivedSocketAddress is small cleanup as we do not need to special case it any more. 

fixes #90125

### CODE REVIEW

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 1]
```diff
@@ -152,6 +152,11 @@ private static unsafe int SysReceive(SafeSocketHandle socket, SocketFlags flags,
                     Count = (UIntPtr)buffer.Length
                 };
 
+                if (socketAddress.Length == 0)
+                {
+                    Debug.Assert(sockAddr == null);
+                }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:158]:
```suggestion
                Debug.Assert(socketAddress.Length != 0 || sockAddr == null);
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 2]
```diff
@@ -484,6 +488,11 @@ private static unsafe int SysReceiveMessageFrom(SafeSocketHandle socket, SocketF
                     Count = (UIntPtr)buffer.Length
                 };
 
+                if (socketAddress.Length == 0)
+                {
+                    Debug.Assert(rawSocketAddress == null);
+                }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:494]:
```suggestion
                Debug.Assert(socketAddress.Length != 0 || rawSocketAddress == null);
```

**Review by stephentoub — APPROVED** [review]

**Comment by wfurt** [comment]
outerloop test failures are unrelated.

**Pull request merged**